src/tool_operate.c

```diff
@@ -675,7 +677,48 @@ static CURLcode post_per_transfer(struct GlobalConfig *global,
         if(per->retry_sleep > RETRY_SLEEP_MAX)
           per->retry_sleep = RETRY_SLEEP_MAX;
       }
-      if(outs->bytes && outs->filename && outs->stream) {
+
+      /* Skip truncation of outfile if auto-resume is enabled for download and
+         the partially received data is good. Currently this is only for HTTP
+         in limited circumstances. */
+      if(config->use_resume && config->resume_from_current &&
+         config->resume_from >= 0 && outs->init == config->resume_from &&
+         !config->use_ascii && !per->uploadfile &&
+         result != CURLE_WRITE_ERROR && result != CURLE_RANGE_ERROR &&
+         outs->bytes > 0 && outs->filename && outs->stream) {
+        const char *scheme;
+        long response = 0;
+
+        curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response);
+        curl_easy_getinfo(curl, CURLINFO_SCHEME, &scheme);
+        scheme = proto_token(scheme);
+
+        if((scheme == proto_http || scheme == proto_https) &&
+           response == (config->resume_from ? 206 : 200)) {
+          notef(config->global,
+                "Keeping %" CURL_FORMAT_CURL_OFF_T " bytes",
+                outs->bytes);
+          if(fflush(outs->stream)) {
+            errorf(config->global, "Failed to flush output file stream");
+            return CURLE_WRITE_ERROR;
+          }
+          if(outs->bytes >= CURL_OFF_T_MAX - outs->init) {
+            errorf(config->global, "Exceeded maximum supported file size ("
+                                   "%" CURL_FORMAT_CURL_OFF_T " + "
+                                   "%" CURL_FORMAT_CURL_OFF_T ")",
+                                   outs->init, outs->bytes);
+            return CURLE_WRITE_ERROR;
+          }
+          truncate = FALSE;
+          outs->init += outs->bytes;
+          outs->bytes = 0;
+          config->resume_from = outs->init;
+          curl_easy_setopt(curl, CURLOPT_RESUME_FROM_LARGE,
+                           config->resume_from);
```

Comment:
Isn't there a chance/risk that this option might get set in a retry scenario due to all the conditions being correct, but then the next attempt fails again and then in a second retry round the conditions are different and this line is not executed?

If so, the `CURLOPT_RESUME_FROM_LARGE` value set in the previous round will remain set, which then probably is wrong?

Reply:
> If so, the `CURLOPT_RESUME_FROM_LARGE` value set in the previous round will remain set, which then probably is wrong?

I updated outs->init and config->resume_from so let's say your scenario happens then it will truncate from the  the updated resume from position which is still correct. The file is truncated to outs->init bytes. truncation can still happen if the conditions are not met.